File: Sudoku.kt
fun main(args: Array<String>) {
    while (true) {
        println(Sudoku.solve(readLine() ?: return))
    }
}

class Sudoku {

    companion object {

        const val UNIT  = 3
        const val GRD   = 0
        const val ROW   = 1
        const val COL   = 2
        const val BOX   = 3
        const val VIEW  = 4
        const val SIZE  = UNIT * UNIT
        const val COORD = SIZE * SIZE * SIZE
        const val BLOCK = VIEW * SIZE * SIZE

        const val OPEN   = 0
        const val FIXED  = 1
        const val BANNED = 2

        const val DONE = SIZE + 1

        const val DIGITS = "123456789"

        fun coord(i: Int, j: Int, k: Int): Int = (i * SIZE + j) * SIZE + k

        fun block(v: Int, p: Int, q: Int): Int = (v * SIZE + p) * SIZE + q

        val parents = Array(COORD) { mutableListOf<Int>() } .apply {
            for (i in 0 until SIZE) {
                for (j in 0 until SIZE) {
                    val p = i / UNIT * UNIT + j / UNIT
                    for (k in 0 until SIZE) {
                        get(coord(i, j, k)).apply {
                            add(block(GRD, i, j))
                            add(block(ROW, i, k))
                            add(block(COL, j, k))
                            add(block(BOX, p, k))
                        }
                    }
                }
            }
        }

        val children = Array(BLOCK) { mutableListOf<Int>() } .apply {
            for (c in 0 until COORD) {
                for (b in parents[c]) {
                    get(b).add(c)
                }
            }
        }

        fun solve(input: String): String {
            return Sudoku().read(input).search()?.show() ?: "NO SOLUTION"
        }
    }

    val state: IntArray
    val count: IntArray
    val queue: MutableList<Int>

    constructor() {
        state = IntArray(COORD) { OPEN }
        count = IntArray(BLOCK) { SIZE }
        queue = mutableListOf<Int>()
    }

    constructor(that: Sudoku) {
        state = that.state.clone()
        count = that.count.clone()
        queue = mutableListOf<Int>()
    }

    fun fix(c: Int): Sudoku {
        queue.add(c)
        return this
    }

    fun read(input: String): Sudoku {
        val len = input.length
        for (i in 0 until SIZE) {
            for (j in 0 until SIZE) {
                if (i * SIZE + j < len) {
                    val k = DIGITS.indexOf(input[i * SIZE + j])
                    if (k != -1) {
                        fix(coord(i, j, k))
                    }
                }
            }
        }
        return this
    }

    fun show(): String {
        val output = CharArray(SIZE * SIZE) { '.' }
        for (i in 0 until SIZE) {
            for (j in 0 until SIZE) {
                for (k in 0 until SIZE) {
                    if (state[coord(i, j, k)] == FIXED) {
                        output[i * SIZE + j] = DIGITS[k]
                    }
                }
            }
        }
        return String(output)
    }

    fun search(): Sudoku? {

        while (!queue.isEmpty()) {
            val c0 = queue.removeAt(0)
            if (state[c0] == FIXED) {
                continue
            }
            if (state[c0] == BANNED) {
                return null
            }
            state[c0] = FIXED
            for (b1 in parents[c0]) {
                count[b1] = DONE
                for (c2 in children[b1]) {
                    if (c2 != c0 && state[c2] == OPEN) {
                        state[c2] = BANNED
                        for (b3 in parents[c2]) {
                            if (b3 != b1) {
                                count[b3]--
                                if (count[b3] == 0) {
                                    return null
                                }
                                if (count[b3] == 1) {
                                    for (c4 in children[b3]) {
                                        if (state[c4] == OPEN) {
                                            fix(c4)
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        val m = count.min()!!
        if (m == DONE) {
            return this
        }

        val b = count.indexOf(m)
        for (c in children[b]) {
            if (state[c] == OPEN) {
                Sudoku(this).fix(c).search()?.let {
                    return it
                }
            }
        }

        return null
    }
}

File: Sudoku.js
'use strict';

const VIEW  = { GRD: 0, ROW: 1, COL: 2, BOX: 3 };
const STATE = { OPEN: 0, FIXED: 1, BANNED: 2 };

const UNIT  = 3;
const SIZE  = UNIT * UNIT;
const COORD = SIZE * SIZE * SIZE;
const BLOCK = Object.keys(VIEW).length * SIZE * SIZE;
const DONE  = SIZE + 1;

const DIGITS = '123456789';

const NO_SOLUTION = 'NO SOLUTION';

const coord = (i, j, k) => (i * SIZE + j) * SIZE + k;
const block = (v, p, q) => (v * SIZE + p) * SIZE + q;

const parents = Array(COORD);
for (let i = 0; i < SIZE; i++) {
  for (let j = 0; j < SIZE; j++) {
    const p = Math.floor(i / UNIT) * UNIT + Math.floor(j / UNIT);
    for (let k = 0; k < SIZE; k++)
      parents[coord(i, j, k)] = [
        block(VIEW.GRD, i, j),
        block(VIEW.ROW, i, k),
        block(VIEW.COL, j, k),
        block(VIEW.BOX, p, k),
      ];
  }
}

const children = Array(BLOCK);
for (let b = 0; b < BLOCK; b++)
  children[b] = Array();
for (let c = 0; c < COORD; c++)
  for (const b of parents[c])
    children[b].push(c);

class Sudoku {

  constructor(that) {
    if (that) {
      this.state = that.state.slice();
      this.count = that.count.slice();
    }
    else {
      this.state = Array(COORD).fill(STATE.OPEN);
      this.count = Array(BLOCK).fill(SIZE);
    }
    this.queue = [];
  }

  enqueue(c) {
    this.queue.push(c);
    return this;
  }

  read(input) {
    for (let i = 0; i < SIZE; i++) {
      for (let j = 0; j < SIZE; j++) {
        if (i * SIZE + j < input.length) {
          const k = DIGITS.indexOf(input[i * SIZE + j]);
          if (k !== -1)
            this.enqueue(coord(i, j, k));
        }
      }
    }
    return this;
  }

  show() {
    const output = Array(SIZE * SIZE).fill('.');
    for (let i = 0; i < SIZE; i++)
      for (let j = 0; j < SIZE; j++)
        for (let k = 0; k < SIZE; k++)
          if (this.state[coord(i, j, k)] === STATE.FIXED)
            output[i * SIZE + j] = DIGITS[k];
    return output.join('');
  }

  search() {

    while (this.queue.length > 0) {
      const c = this.queue.shift();
      switch (this.state[c]) {
        case STATE.OPEN:
          this.fix(c);
          break;
        case STATE.FIXED:
          break;
        case STATE.BANNED:
          throw NO_SOLUTION;
      }
    }

    const min = Math.min.apply(null, this.count);
    if (min === DONE)
      return this;

    const b = this.count.indexOf(min);
    for (const c of children[b]) {
      if (this.state[c] === STATE.OPEN) {
        try {
          return new Sudoku(this).enqueue(c).search();
        }
        catch (e) {
          if (e !== NO_SOLUTION)
            throw e;
        }
      }
    }

    throw NO_SOLUTION;
  }

  fix(c) {
      this.state[c] = STATE.FIXED;
      for (const b of parents[c])
        this.mark(b);
  }

  mark(b) {
    this.count[b] = DONE;
    for (const c of children[b])
      if (this.state[c] == STATE.OPEN)
        this.ban(c);
  }

  ban(c) {
    this.state[c] = STATE.BANNED;
    for (const b of parents[c])
      if (this.count[b] != DONE)
        this.countdown(b);
  }

  countdown(b) {
    this.count[b]--;
    if (this.count[b] == 0)
      throw NO_SOLUTION;
    if (this.count[b] == 1)
      for (const c of children[b])
        if (this.state[c] == STATE.OPEN)
          this.enqueue(c);
  }
}

function solve(input) {
  try {
    return new Sudoku().read(input).search().show();
  }
  catch (e) {
    if (e !== NO_SOLUTION)
      throw e;
    return e;
  }
}

require('readline')
  .createInterface({ input: process.stdin, terminal: false })
  .on('line', input => console.log(solve(input)));
